File: src/claude_memory_kit/tools/recall.py
import asyncio
import logging

from ..store import Store

log = logging.getLogger("cmk")


async def do_recall(
    store: Store, query: str, user_id: str = "local"
) -> str:
    results = []
    seen_ids: set[str] = set()

    # 1. Hybrid search (dense + sparse with RRF fusion) via Qdrant
    try:
        vec_results = await asyncio.to_thread(
            store.qdrant.search, query, 10, user_id
        )
        for mem_id, score in vec_results:
            if mem_id not in seen_ids:
                seen_ids.add(mem_id)
                full = store.qdrant.get_memory(mem_id, user_id=user_id)
                if full:
                    store.qdrant.touch_memory(mem_id, user_id=user_id)
                    person = full.person or "?"
                    results.append(
                        f"[{full.gate.value}, score={score:.2f}] "
                        f"({full.created:%Y-%m-%d}, {person}) "
                        f"{full.content}\n  id: {full.id}"
                    )
    except Exception as e:
        log.warning("hybrid search failed: %s", e)

    # 2. Text search fallback when hybrid returned nothing
    if not results and not store.qdrant._disabled:
        try:
            text_results = await asyncio.to_thread(
                store.qdrant.search_text, query, 5, user_id
            )
            for mem_id, score in text_results:
                if mem_id not in seen_ids:
                    seen_ids.add(mem_id)
                    full = store.qdrant.get_memory(mem_id, user_id=user_id)
                    if full:
                        store.qdrant.touch_memory(mem_id, user_id=user_id)
                        person = full.person or "?"
                        results.append(
                            f"[{full.gate.value}, text] "
                            f"({full.created:%Y-%m-%d}, {person}) "
                            f"{full.content}\n  id: {full.id}"
                        )
        except Exception as e:
            log.warning("text search failed: %s", e)

    # 3. Graph traversal for sparse results
    if len(results) < 3:
        for mid in list(seen_ids)[:2]:
            related = store.qdrant.find_related(
                mid, depth=2, user_id=user_id
            )
            for rel in related:
                rid = rel["id"]
                if rid not in seen_ids:
                    seen_ids.add(rid)
                    preview = rel.get("content", "")[:80]
                    results.append(
                        f"[graph: {rel['relation']}] "
                        f"{preview} (id: {rid})"
                    )

    if not results:
        return "No memories found matching that query."

    return f"Found {len(results)} memories:\n\n" + "\n\n".join(results)

File: src/tools/recall.rs
use anyhow::Result;
use crate::server::MemoryServer;
use crate::store::markdown;

impl MemoryServer {
    pub async fn do_recall(&self, query: &str) -> Result<String> {
        let store = self.store.lock().await;
        let mut results = Vec::new();
        let mut seen_ids = std::collections::HashSet::new();

        // 1. FTS5 search
        match store.db.search_fts(query, 5) {
            Ok(fts_results) => {
                for mem in fts_results {
                    if seen_ids.insert(mem.id.clone()) {
                        let _ = store.db.touch_memory(&mem.id);
                        results.push(format!(
                            "[{}] ({}, {}) {}\n  id: {}",
                            mem.gate.as_str(),
                            mem.created.format("%Y-%m-%d"),
                            mem.person.as_deref().unwrap_or("?"),
                            mem.content,
                            mem.id,
                        ));
                    }
                }
            }
            Err(e) => tracing::warn!("fts5 search failed: {}", e),
        }

        // 2. Qdrant vector search
        if let Some(ref vectors) = store.vectors {
            match vectors.search_similar(query, 5).await {
                Ok(vec_results) => {
                    for (mem_id, score) in vec_results {
                        if seen_ids.insert(mem_id.clone()) {
                            results.push(format!(
                                "[vector match, score={:.2}] id: {}", score, mem_id
                            ));
                        }
                    }
                }
                Err(e) => tracing::warn!("qdrant search failed: {}", e),
            }
        }

        // 3. Neo4j graph traversal (for sparse results)
        if results.len() < 3 {
            if let Some(ref graph) = store.graph {
                for id in seen_ids.clone().iter().take(2) {
                    match graph.find_related(id, 2).await {
                        Ok(related) => {
                            for (rid, relation, preview) in related {
                                if seen_ids.insert(rid.clone()) {
                                    results.push(format!(
                                        "[graph: {}] {} (id: {})",
                                        relation, preview, rid
                                    ));
                                }
                            }
                        }
                        Err(e) => tracing::warn!("neo4j traversal failed: {}", e),
                    }
                }
            }
        }

        // 4. Fallback: grep markdown
        if results.is_empty() {
            match markdown::search_all(&self.store_path, query) {
                Ok(grep_results) => {
                    for content in grep_results.iter().take(3) {
                        let preview: String = content.chars().take(300).collect();
                        results.push(format!("[file search] {}", preview));
                    }
                }
                Err(e) => tracing::warn!("markdown search failed: {}", e),
            }
        }

        if results.is_empty() {
            Ok("No memories found matching that query.".into())
        } else {
            Ok(format!(
                "Found {} memories:\n\n{}",
                results.len(),
                results.join("\n\n")
            ))
        }
    }
}
